basic_module/distribution/src/lib.rs

```diff
@@ -13,3 +13,305 @@
+    fn next(&mut self) -> Option<(&'a Public, u64)> {
+        if let Some((stakeholder, stake)) = self.stake_holdings.next() {
+            let share = share(self.total_stakes, *stake, self.total_min_fee);
+            self.remaining_fee = self.remaining_fee.checked_sub(share).expect("Remaining fee shouldn't be depleted");
+            Some((stakeholder, share))
+        } else {
+            None
+        }
+    }
+}
+
+fn share(total_stakes: u64, stake: u64, total_min_fee: u64) -> u64 {
+    assert!(total_stakes >= stake);
+    u64::try_from((u128::from(total_min_fee) * u128::from(stake)) / u128::from(total_stakes)).unwrap()
+}
+
+pub fn get_context() -> &'static mut dyn Context {
+    // This function should be implemented after the context has been formatted.
+    unimplemented!();
+}
+
+pub fn get_storage_id() -> StorageId {
+    // This function should be implemented after the storageId has been formatted.
+    unimplemented!()
+}
+
+pub fn get_key() -> H256 {
+    // This function should be implemented after the storageId has been formatted.
+    H256::default()
+}
+
+pub fn get_intermediate_rewards(_key: H256) -> IntermediateRewards {
+    get_context().get(get_storage_id(), &get_key()).map(|vec| vec.into()).unwrap_or_default()
+}
+
+/// reward = floor(intermediate_rewards * (a * number_of_signatures / number_of_blocks_in_term + b) / 10)
+pub fn final_rewards(intermediate_reward: u64, number_of_signatures: u64, number_of_blocks_in_term: u64) -> u64 {
+    let (a, b) = if number_of_signatures * 3 >= number_of_blocks_in_term * 2 {
+        // number_of_signatures / number_of_blocks_in_term >= 2 / 3
+        // x * 3/10 + 7/10
+        (3, 7)
+    } else if number_of_signatures * 2 >= number_of_blocks_in_term {
+        // number_of_signatures / number_of_blocks_in_term >= 1 / 2
+        // x * 48/10 - 23/10
+        (48, -23)
+    } else if number_of_signatures * 3 >= number_of_blocks_in_term {
+        // number_of_signatures / number_of_blocks_in_term >= 1 / 3
+        // x * 6/10 - 2/10
+        (6, -2)
+    } else {
+        // 1 / 3 > number_of_signatures / number_of_blocks_in_term
+        // 0
+        assert!(
+            number_of_blocks_in_term > 3 * number_of_signatures,
+            "number_of_signatures / number_of_blocks_in_term = {}",
+            (number_of_signatures as f64) / (number_of_blocks_in_term as f64)
+        );
+        (0, 0)
+    };
+    let numerator = i128::from(intermediate_reward)
+        * (a * i128::from(number_of_signatures) + b * i128::from(number_of_blocks_in_term));
+    assert!(numerator >= 0);
+    let denominator = 10 * i128::from(number_of_blocks_in_term);
+    // Rust's division rounds towards zero.
+    u64::try_from(numerator / denominator).unwrap()
+}
+
+#[allow(dead_code)]
+#[derive(Default)]
+struct WorkInfo {
+    proposed: usize,
+    missed: usize,
+    signed: u64,
+}
+
+fn aggregate_work_info(start_of_the_next_term_header: Header) -> Result<HashMap<Public, WorkInfo>, Error> {
```

Comment:
I think this function can not expect any errors. Is the function signature intentional ?